src/lib/ast.rs

```diff
@@ -0,0 +1,617 @@
+/// these information that the `lrpar` crate will provide, and represent the
+/// type (e.g. `expr`, `factor`, `term`) and information (e.g. literals) stored
+/// within terminals and non-terminals in a given grammar. In most uses, `T` and
+/// `U` will both be `String` types.
+pub struct Arena<T, U> {
+    nodes: Vec<Node<T, U>>,
+}
+
+impl<T, U> Index<NodeId> for Arena<T, U> {
+    type Output = Node<T, U>;
+
+    fn index(&self, node: NodeId) -> &Node<T, U> {
+        &self.nodes[node]
+    }
+}
+
+impl<T, U> IndexMut<NodeId> for Arena<T, U> {
+    fn index_mut(&mut self, node: NodeId) -> &mut Node<T, U> {
+        &mut self.nodes[node]
+    }
+}
+
+impl<T, U> Arena<T, U> {
+    /// Create an empty `Arena`.
+    pub fn new() -> Arena<T, U> {
+        Arena { nodes: Vec::new() }
+    }
+
+    /// Create a new node from its data.
+    pub fn new_node(&mut self, data: T, ty: U, indent: u16) -> NodeId {
+        let next_index = self.nodes.len();
+        let node = Node::new(data, ty, indent);
+        self.nodes.push(node);
+        next_index
+    }
+
+    /// Return `true` if the arena is empty, `false` otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.nodes.is_empty()
+    }
+
+    /// Make one node the child of another.
+    /// Panic if either `NodeId` is not found.
+    pub fn make_child_of(&mut self, child: NodeId, parent: NodeId) {
+        if child >= self.nodes.len() {
+            panic!("Node ID {} not found in arena.", child);
+        }
+        if parent >= self.nodes.len() {
+            panic!("Node ID {} not found in arena.", parent);
+        }
+        self[child].parent = Some(parent);
+        match self[parent].first_child {
+            None => self[parent].first_child = Some(child),
+            Some(_) => (),
+        };
+        match self[parent].last_child {
+            None => self[parent].last_child = Some(child),
+            Some(id) => {
+                self[parent].last_child = Some(child);
+                self[id].next_sibling = Some(child);
+                self[child].previous_sibling = Some(id);
+            }
+        };
+    }
+
+    // Return a vector of edges between nodes in this arena.
+    // Each edge pair `(n1, n2)` denotes an edge between nodes `n1` and `n2`,
+    // both numbers being `NodeId.index` indices within this arena.
+    fn get_edges(&self) -> Vec<EdgeId> {
+        let mut edges: Vec<EdgeId> = vec![];
+        for index in 0..self.nodes.len() {
+            let node = &self.nodes[index];
+            match node.parent {
+                None => (),
+                Some(par) => edges.push((par, index)),
+            }
+        }
+        edges
+    }
+}
+
+impl Arena<String, String> {
+    /// Create a graphviz representation of this arena and write to buffer.
+    pub fn render_dotgraph<W: Write>(&self, buffer: &mut W) -> Result<(), io::Error> {
+        dot::render(self, buffer)
+    }
+}
+
+impl<'a> dot::Labeller<'a, NodeId, EdgeId> for Arena<String, String> {
```

Comment:
I wonder -- and this may be a bad idea! -- if the graphviz stuff should be in a new module? That way you would probably feel more comfortable saying `use dot::{Labeller, ...}` and then not having to qualify `dot::` everywhere.